File: frontend/src/domain/guides.js
/**
 * The photographer's guide, in frame units (0-1), independent of the cell grid.
 *
 * The cell mesh (A1..G9) is how the model points at things. A guide is how a
 * photographer reads a frame — thirds, the phi grid, the diagonal method, a
 * centre axis — and what they want to know is whether the frame is sitting on
 * it. So these return lines and landing points, and `offset` says how far the
 * subject is from the nearest one.
 *
 * Which guide is chosen lives in the backend (`domain/guides.py`) and arrives
 * on `analysis.composition.guide`; this file only knows how to draw them.
 */

const PHI = 0.382 // 1 : 0.618 : 1

export const GUIDES = ['thirds', 'phi', 'diagonals', 'centre', 'fill', 'none']

export const GUIDE_LABELS = {
  thirds: 'thirds',
  phi: 'phi',
  diagonals: 'diagonals',
  centre: 'centre',
  fill: 'frame',
  none: 'off',
}

/**
 * Lines and power points for a guide, in frame units. `aspect` is width/height
 * in pixels: the diagonal method's corner bisectors are 45° on the print, not
 * 45° in normalised space, so they need it.
 */
export function geometry(guide, aspect = 1) {
  if (guide === 'centre') {
    return {
      lines: [
        { x1: 0.5, y1: 0, x2: 0.5, y2: 1 },
        { x1: 0, y1: 0.5, x2: 1, y2: 0.5 },
      ],
      points: [{ x: 0.5, y: 0.5 }],
    }
  }

  if (guide === 'diagonals') {
    // The diagonal method: the two diagonals, plus a 45° bisector from each
    // corner, which is where a subject placed on a diagonal wants to sit.
    const dx = Math.min(1, 1 / aspect)
    const dy = Math.min(1, aspect)
    return {
      lines: [
        { x1: 0, y1: 0, x2: 1, y2: 1 },
        { x1: 1, y1: 0, x2: 0, y2: 1 },
        { x1: 0, y1: 0, x2: dx, y2: dy },
        { x1: 1, y1: 0, x2: 1 - dx, y2: dy },
        { x1: 0, y1: 1, x2: dx, y2: 1 - dy },
        { x1: 1, y1: 1, x2: 1 - dx, y2: 1 - dy },
      ],
      points: [],
    }
  }

  if (guide === 'thirds' || guide === 'phi') {
    const at = guide === 'phi' ? [PHI, 1 - PHI] : [1 / 3, 2 / 3]
    const lines = []
    const points = []
    for (const f of at) {
      lines.push({ x1: f, y1: 0, x2: f, y2: 1 })
      lines.push({ x1: 0, y1: f, x2: 1, y2: f })
    }
    for (const x of at) for (const y of at) points.push({ x, y })
    return { lines, points }
  }

  return { lines: [], points: [] }
}

/** Nearest landing point, and how far the subject sits from it. */
export function offset(guide, point) {
  const { points } = geometry(guide)
  if (!point || !points.length) return null
  let best = null
  for (const candidate of points) {
    const distance = Math.hypot(candidate.x - point.x, candidate.y - point.y)
    if (!best || distance < best.distance) best = { point: candidate, distance }
  }
  return best
}

/**
 * What to tell the photographer about the fit — but only when the read is
 * precise enough to say it. Cells quantise to a seventh of the width, so
 * without a sub-cell subject point there is no honest number to print.
 */
export function verdict(guide, point, { cellWidth = 1 / 7 } = {}) {
  if (guide === 'none' || !point) return ''
  const near = offset(guide, point)
  if (!near) return ''
  const percent = Math.round(near.distance * 100)
  if (near.distance < cellWidth / 2) return `sitting on the ${label(guide)}`
  return `${percent}% of the frame off the nearest ${label(guide)}`
}

function label(guide) {
  if (guide === 'centre') return 'centre'
  if (guide === 'phi') return 'phi point'
  return 'thirds point'
}

File: backend/app/domain/guides.py
"""Which compositional guide to draw over a frame, for a human.

Two different grids serve two different readers. The cell grid (`A1`..`G9`)
is an addressing system: it exists so a model can point at things and so code
can map that back to pixels. It is not a compositional idea and showing it to
a photographer teaches nothing.

What a photographer reads is a guide — thirds, the phi grid, the diagonal
method, a centre axis — and what they want to know is whether the frame is
sitting on it. So the guide is chosen from the strongest retained spatial
Technique, and the app draws that one, not the mesh.

Pure. The renderers (SVG in the browser, PIL for the Drive review) own the
geometry; this owns only the choice.
"""

from app.domain.entities import TechniqueEvidence

#: Guide ids the renderers know how to draw.
THIRDS = "thirds"
CENTRE = "centre"
DIAGONALS = "diagonals"
FILL = "fill"
PHI = "phi"
NONE = "none"

GUIDES = (THIRDS, CENTRE, DIAGONALS, FILL, PHI, NONE)

#: Only composition techniques choose a guide; light and colour have no
#: geometry to draw. A technique missing here simply does not vote.
BY_TECHNIQUE = {
    "rule_of_thirds": THIRDS,
    "rule_of_odds": THIRDS,
    "horizon_placement": THIRDS,
    "negative_space": THIRDS,
    "centre_composition": CENTRE,
    "symmetry": CENTRE,
    "reflections": CENTRE,
    "leading_lines": DIAGONALS,
    "diagonals": DIAGONALS,
    "layering": DIAGONALS,
    "fill_the_frame": FILL,
    "minimalism": THIRDS,
}

#: With no retained spatial Technique, thirds is the guide every photographer already
#: has in their head — drawn dimmed, as a reference rather than a claim.
FALLBACK = THIRDS

#: The two grids a subject can be placed on, as fractions of the frame. They sit
#: 0.049 apart, which no lens can tell apart by eye — so the choice between them
#: is measured here from the point the Composer gave, not asked of a model.
THIRDS_LINES: tuple[float, ...] = (1 / 3, 2 / 3)
PHI_LINES: tuple[float, ...] = (0.382, 0.618)


def _distance(position: float, lines: tuple[float, ...]) -> float:
    return min(abs(line - position) for line in lines)


def refine(guide: str, subject_x: float | None, subject_y: float | None) -> str:
    """Thirds or phi, decided by where the subject actually is.

    Only the thirds guide is refined. A frame the panel read as centred or
    diagonal is not a placement question, and redrawing it on a phi grid would
    answer something nobody asked."""
    if guide != THIRDS:
        return guide
    positions = [value for value in (subject_x, subject_y) if value is not None]
    if not positions:
        return guide
    thirds = min(_distance(value, THIRDS_LINES) for value in positions)
    phi = min(_distance(value, PHI_LINES) for value in positions)
    return PHI if phi < thirds else THIRDS


def choose(
    techniques: list[TechniqueEvidence],
    subject_x: float | None = None,
    subject_y: float | None = None,
) -> str:
    """The guide for the strongest retained spatial Technique,
    refined to the grid the subject is actually nearer."""
    best: tuple[float, str] | None = None
    for evidence in techniques:
        guide = BY_TECHNIQUE.get(evidence.technique_id)
        if guide is None:
            continue
        if best is None or evidence.confidence > best[0]:
            best = (evidence.confidence, guide)
    return refine(best[1] if best else FALLBACK, subject_x, subject_y)
